crit/mempages.go

```diff
@@ -193,3 +194,72 @@ func (mr *MemoryReader) GetShmemSize() (int64, error) {
 
 	return size, nil
 }
+
+// PatternMatch represents a match when searching for a pattern in memory.
+type PatternMatch struct {
+	Vaddr   uint64
+	Length  int
+	Context int
+	Pattern string
+}
+
+// SearchPattern searches for a pattern in the process memory pages.
+// If the pattern is found, it returns slices of strings containing the matched patterns
+// along with the specified context (number of bytes before and after each match).
+func (mr *MemoryReader) SearchPattern(pattern string, context int) ([]PatternMatch, error) {
+	if context < 0 {
+		return nil, errors.New("context size cannot be negative")
+	}
+
+	regexPattern, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []PatternMatch
+
+	for _, entry := range mr.pagemapEntries {
+		startAddr := entry.GetVaddr()
+		endAddr := startAddr + uint64(entry.GetNrPages())*uint64(mr.pageSize)
+
+		// Read the entire pages content
+		buffer, err := mr.GetMemPages(startAddr, endAddr)
+		if err != nil {
```

Comment:
@behouba What would be the performance of the search functionality if we have a container checkpoint that has pages file larger than 100 GB and checkpointctl is running on a laptop with 16 GB memory?

Reply:
I guess this laptop will run out of memory because the function will try to read the entire 100GB at once. 
Thank you @rst0git for pointing this out. I will work on an optimization alternative.

Reply:
> I guess this laptop will run out of memory because the function will try to read the entire 100GB at once.

I'm not sure. It looks like [`getPage()`](https://github.com/checkpoint-restore/go-criu/blob/972bd31ea11289771b4b6cb41e80fbd2f9425d7e/crit/mempages.go#L117) reads one page at a time. However, if we have a checkpoint with 100 GB pages file (≈ 26,214,400 pages), the search functionality might be very slow.

Reply:
@rst0git you were right. The previous implementation had a serious performance issue due to repeated calls to [GetMemPage()](https://github.com/checkpoint-restore/go-criu/blob/972bd31ea11289771b4b6cb41e80fbd2f9425d7e/crit/mempages.go#L69).  In my local tests with a 100MB checkpoint, the search functionality took an average of 20 seconds to complete. I optimized the implementation, and the new version is faster. Can you please take a look?

I have also opened the draft PR for checkpointctl [here](https://github.com/checkpoint-restore/checkpointctl/pull/141).